mindmeld/components/entity_resolver.py

```diff
@@ -542,43 +231,120 @@ def _process_entity_map(entity_type,
+    def _format_entity_map(entity_map):
+        if not entity_map.get("entities", []):
+            return entity_map
+
+        # format (if required) entities key-field
+        new_entities = []
+        id_counter = 0
+        for ent in entity_map.get("entities", []):
+            cname = ent.get("cname", None)
+            if not cname:
+                continue
+            _id, whitelist = ent.get("id", None), ent.get("whitelist", [])
+            if not _id:
+                _id = str(id_counter)
+                id_counter += 1
+            new_entities.append({"id": _id, "cname": cname, "whitelist": whitelist})
+
+        entity_map["entities"] = new_entities
+        return entity_map
+
+    def _trim_results(self, results, top_n):
+        """
+        Trims down the results generated by any ER class, finally populating at max top_n documents
 
         Args:
-            clean (bool): If ``True``, deletes and recreates the index from scratch instead of
-                            updating the existing index with synonyms in the mapping.json.
-            entity_map (json): json data loaded from `mapping.json` file for the entity type
+            results (list[dict]): Each element in this list is a result dictions with keys such as
+                `id` (optional), `cname`, `score` and any others
+            top_n (int): Number of top documents required to be populated
+
+        Returns:
+            list[dict]: if trimmed, a list similar to `results` but with fewer elements,
+                        else, the `results` list as-is is returned
         """
-        raise NotImplementedError
 
-    def fit(self, clean=False):
+        if not results:
+            return None
+
+        if not isinstance(top_n, int) or top_n <= 0:
+            msg = f"The value of `top_n` set to `{top_n}` during predictions in " \
+                  f"{self.__class__.__name__}. This will result in an unsorted list of documents. "
+            logger.warning(msg)
+            return results
+
+        # Obtain top scored for each doc id, if scores field exist in results
+        best_results = {}
+        for result in results:
+            if "score" not in result:
+                return results
+            _id = result["id"] if "id" in result else result["cname"]
+            if _id not in best_results or result["score"] > best_results[_id]["score"]:
+                best_results[_id] = result
+
+        # sort as per requirement using the score field
+        results = [*best_results.values()]
+        n_scores = len(results)
+        if n_scores < top_n:
+            msg = f"Retrieved only {len(results)} entity resolutions instead of asked " \
+                  f"number {top_n} for entity type {self.type}"
+            logger.info(msg)
+            return results
+        #  n_scores >= top_n
+        _sim_scores = np.asarray([val["score"] for val in results])
+        _top_inds = _sim_scores.argpartition(n_scores - top_n)[-top_n:]
```

Comment:
Its hard to parse this operation. If one wants to sort top_n, I would think we could do this:

```
_top_inds = _sim_scores.partition(top_n)

or 

_top_inds = _sim_scores[_sim_scores.argpartition(top_n)]
```

Reply:
`top_n` (integer) here actually means number of top items to be collected. Not an array! I think it make sense to change this variable to `size` to avoid confusion. 

Because the partition and argpartition funcs move smaller values to left side of array index=top_n-1, I move all scores less than top_n to left side and retain all top_n scores from array index top_n to end of array. And then simply collect the last top_n indices of array. Let me know if it makes sense; else I'll breakdown the op.